src/Main.elm

```diff
@@ -1,10 +1,209 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
-import Html exposing (Html)
+import Browser
+import Data.Audience as A
+import Data.AudienceFolder as F
+import Dict
+import Html
+import Html.Attributes as Hat
+import Html.Events as Hev
+import Json.Decode as Jd
+import Parser as P
+import Set
 
 
-main : Html msg
+main : Program () Model Msg
 main =
-    Html.text "There will be app soon!"
+    Browser.sandbox
+        { init = init
+        , view = view
+        , update = update
+        }
+
+
+type Msg
+    = FolderClick Int
+    | GoUpClick
+
+
+type alias Model =
+    Result String GoodModel
+
+
+{-| The integer IDs are not the same as the ones provided in the data.
+It seems likely that the provided IDs are row IDS from the database,
+and that folders and audiences are kept in different tables, so
+the IDs may well conflict. A globally unique ID is needed here.
+
+This model design is inspired by the book Data-Oriented Design by
+Richard Fabian, see <https://www.dataorienteddesign.com/dodbook> .
+It's an idea from game dev that says that using tabular data
+structures gives better performance. I am experimenting with using
+it in high-level languages, like Elm and Go. I am finding so far
+that it gives a really elegant and minimal internal state.
+
+-}
+type alias GoodModel =
+    -- The keys are child IDs and the values are parent IDs. Root
+    -- items are not included. If a map lookup fails then the item
+    -- is root, so there are no internal errors caused by failing map
+    -- lookups. The reason for using a dictionary rather than a list
+    -- is to enforce that a child can have only one parent.
+    { parents : Dict.Dict Int Int
```

Comment:
I considered having something like:

```
type ChildId
    = ChildId Int
```

and 

```
type ParentId
    = ParentId Int
```

Or I could have type aliases instead of wrappers. One problem with the wrappers is that they are not comparable types in Elm (I think), so can't be used directly as Dict keys. One way to deal with this would be to make a module containing a Dict Int Int structure, but only export getters and setters for ChildId and ParentId.

I'm torn on this one. I really like the simplicity of plain Dict Int Int. I think it's also important to bear in mind that indirection has a readability cost, because whenever you see 'ChildId' in the code you have to look up its definition, which is one more bit of cognitive effort. And if it's in a separate module that's even harder. On the other hand, when mapping over the dict, functions will have arguments like ... Int -> Int ... and how can the reader tell which is which? They have to look back to where the function is called. There are the variable names, but they are not checked by the type system.

In general I'm not that keen on type aliases, unless they are for really huge types like record types that are not practical to include in full. They don't improve type safety, and increase mental indirection for the reader.

I am perhaps over-reacting slightly to some Haskell code I have seen that has so many type aliases and newtypes that it's really hard to know what's going on, and what the actual data is underneath. Elm doesn't seem to suffer from this problem so much.

I also feel that in this case it would be hard for the code to be subtly wrong if somewhere the child and parent IDs got mixed up. I think it would be noticeable straight away.

When in doubt, I go on the principle that less code is better.

Reply:
> One problem with the wrappers is that they are not comparable types in Elm (I think), so can't be used directly as Dict keys.

FYI there are packages like [`turboMaCk/any-dict`](https://package.elm-lang.org/packages/turboMaCk/any-dict/latest/) (more performant, needs you to provide a `toComparable` function) and [`pzp1997/assoc-list`](https://package.elm-lang.org/packages/pzp1997/assoc-list/latest/) (less performant, but nicer library API) with which you can have a `Dict customtype a`

Reply:
IMHO even the type aliases would be better (would convey more meaning) than `Dict Int Int` as it could mean anything.

I would go with the custom type wrappers though.

In our real app code we go even further and do it with phantom types:

```elm
type Id a
    = Id String

type AudienceIdTag
    = AudienceIdTag

type alias AudienceId =
    Id AudienceIdTag

type AudienceFolderIdTag
    = AudienceFolderIdTag

type alias AudienceFolderId =
    Id AudienceFolderIdTag
```
which allows us to have some common logic for `Dict (Id tag) ...` and `Set (Id tag)`.

Reply:
This dict is now gone, following your comments below, but one nice thing about using plain ints is that if I choose the IDs carefully, Dict Int Int could be changed to Array Int. So the child IDs are the array positions and the parent IDs are the array contents. I would think that the array lookups would be way faster than the dict lookups, but it's hard to be sure without benchmarking.

Reply:
Well, we can always benchmark!
```elm
module ArrayDictAccess exposing (main)

import Array
import Benchmark exposing (..)
import Benchmark.Runner exposing (BenchmarkProgram, program)
import Dict


main : BenchmarkProgram
main =
    program suite


suite : Benchmark
suite =
    let
        sampleDict =
            List.range 0 10000
                |> List.map (\i -> ( i, 3 * i ))
                |> Dict.fromList

        sampleArray =
            List.range 0 10000
                |> List.map (\i -> 3 * i)
                |> Array.fromList
    in
    Benchmark.compare "Access"
        "Dict"
        (\_ -> Dict.get 5000 sampleDict)
        "Array"
        (\_ -> Array.get 5000 sampleArray)
```
<img width="584" alt="Screenshot 2021-09-15 at 12 01 57" src="https://user-images.githubusercontent.com/149425/133413675-fff167a0-3580-465a-86d3-2f6278ede4d7.png">

Reply:
Wow that looks good! Is that BrianHicks/elm-benchmark you're using? It says it's not compatible with 0.19.

Reply:
@8n8 Correct, it just got moved to [`elm-explorations/benchmark`](https://package.elm-lang.org/packages/elm-explorations/benchmark/latest/)! :)

Reply:
I've just refactored it again, and now all the data is in arrays of basic types, with the array positions acting as IDs. There's a slight sacrifice of type safety, because I couldn't think of a way to guarantee in the types that everything has a name. But all the lookups are fast array lookups instead of dict lookups. I also added a breadcrumb.

Reply:
I just thought of a way to avoid the loss of type safety, by giving the failing name lookups the meaning that the failing ID is the root folder, which doesn't have a name.

Reply:
So now the whole internal state is just:
```
type alias GoodModel =
    { names : Array.Array String
    , parents : Array.Array Int
    , firstAudienceId : Int
    , currentLevel : Int
    }
```